src/Microsoft.AspNet.Razor/SourceLocation.cs

```diff
@@ -89,57 +165,82 @@ public static SourceLocation Add(SourceLocation left, SourceLocation right)
             else
             {
                 return new SourceLocation(
+                    left.FilePath,
                     left.AbsoluteIndex + right.AbsoluteIndex,
                     left.LineIndex + right.LineIndex,
                     left.CharacterIndex + right.CharacterIndex);
             }
         }
 
-        public static SourceLocation Subtract(SourceLocation left, SourceLocation right)
+        /// <summary>
+        /// Subtracts two <see cref="SourceLocation"/>s.
+        /// </summary>
+        /// <param name="left">The left operand.</param>
+        /// <param name="right">The right operand.</param>
+        /// <returns>A <see cref="SourceLocation"/> that is the difference of the left and right operands.</returns>
+        /// <exception cref="ArgumentException">if the <see cref="FilePath"/> of the left and right operands
+        /// are different.</exception>
+        public static SourceLocation operator -(SourceLocation left, SourceLocation right)
         {
-            return new SourceLocation(
-                left.AbsoluteIndex - right.AbsoluteIndex,
-                left.LineIndex - right.LineIndex,
-                left.LineIndex != right.LineIndex ? left.CharacterIndex : left.CharacterIndex - right.CharacterIndex);
-        }
+            if (!string.Equals(left.FilePath, right.FilePath, StringComparison.Ordinal))
+            {
+                throw new ArgumentException(
+                    RazorResources.FormatSourceLocationFilePathDoesNotMatch(nameof(SourceLocation), "-"),
+                    nameof(right));
+            }
 
-        private static SourceLocation CreateUndefined()
-        {
-            var sl = new SourceLocation();
-            sl.AbsoluteIndex = -1;
-            sl.LineIndex = -1;
-            sl.CharacterIndex = -1;
-            return sl;
+            var characterIndex = left.LineIndex != right.LineIndex ?
+                left.CharacterIndex : left.CharacterIndex - right.CharacterIndex;
+
+            return new SourceLocation(
+                filePath: null,
+                absoluteIndex: left.AbsoluteIndex - right.AbsoluteIndex,
+                lineIndex: left.LineIndex - right.LineIndex,
+                characterIndex: characterIndex);
         }
 
+        /// <summary>
+        /// Determines whether the first operand is less than the second operand.
+        /// </summary>
+        /// <param name="left">The left operand.</param>
+        /// <param name="right">The right operand.</param>
+        /// <returns><c>true</c> if <paramref name="left"/> is less than <paramref name="right"/>.</returns>
```

Comment:
suggest duplicating the `<exception/>` documentation from `CompareTo()` and `Equals()` in the methods that use them